src/server/v2.0/handler/artifact.go

```diff
@@ -16,32 +16,157 @@ package handler
+func newArtifactAPI() *artifactAPI {
+	return &artifactAPI{
+		artCtl:  artifact.Ctl,
+		proMgr:  project.Mgr,
+		repoMgr: repository.Mgr,
+	}
+}
+
+type artifactAPI struct {
 	BaseAPI
+	artCtl  artifact.Controller
+	proMgr  project.Manager
+	repoMgr repository.Manager
 }
 
-// DeleteArtifact ...
-func (api *ArtifactAPI) DeleteArtifact(ctx context.Context, params operation.DeleteArtifactParams) middleware.Responder {
-	return operation.NewDeleteArtifactOK()
+// TODO do auth in a separate middleware
+
+func (a *artifactAPI) List(ctx context.Context, params operation.ListParams) middleware.Responder {
+	// set query
+	query := &q.Query{
+		Keywords: map[string]interface{}{},
+	}
+	if params.Type != nil {
+		query.Keywords["Type"] = *(params.Type)
+	}
+	if params.Page != nil {
+		query.PageNumber = *(params.Page)
+	}
+	if params.PageSize != nil {
+		query.PageSize = *(params.PageSize)
+	}
+	repository, err := a.repoMgr.GetByName(ctx, fmt.Sprintf("%s/%s", params.ProjectName, params.RepositoryName))
+	if err != nil {
+		return a.SendError(ctx, err)
+	}
+	query.Keywords["RepositoryID"] = repository.RepositoryID
+
+	// set option
+	option := &artifact.Option{
+		WithTag: true, // return the tag by default
+	}
+	if params.WithTag != nil {
+		option.WithTag = *(params.WithTag)
+	}
+	if option.WithTag {
+		if params.WithImmutableStatus != nil {
+			option.TagOption = &artifact.TagOption{
+				WithImmutableStatus: *(params.WithImmutableStatus),
+			}
+		}
+	}
+	if params.WithLabel != nil {
+		option.WithLabel = *(params.WithLabel)
+	}
+	if params.WithScanReport != nil {
+		option.WithScanReport = *(params.WithScanReport)
+	}
+	if params.WithSignatrue != nil {
+		option.WithSignature = *(params.WithSignatrue)
+	}
+
+	// list artifacts according to the query and option
+	total, arts, err := a.artCtl.List(ctx, query, option)
+	if err != nil {
+		return a.SendError(ctx, err)
+	}
+
+	var artifacts []*models.Artifact
+	for _, art := range arts {
+		artifacts = append(artifacts, convert(art))
+	}
+
+	// TODO add link header
+	return operation.NewListOK().WithXTotalCount(total).WithLink("").WithPayload(artifacts)
 }
 
-// ListArtifacts ...
-func (api *ArtifactAPI) ListArtifacts(ctx context.Context, params operation.ListArtifactsParams) middleware.Responder {
-	return operation.NewListArtifactsOK()
+func (a *artifactAPI) Get(ctx context.Context, params operation.GetParams) middleware.Responder {
+	// TODO implement
+	return operation.NewGetOK()
 }
 
-// ReadArtifact ...
-func (api *ArtifactAPI) ReadArtifact(ctx context.Context, params operation.ReadArtifactParams) middleware.Responder {
-	return operation.NewReadArtifactOK()
+func (a *artifactAPI) Delete(ctx context.Context, params operation.DeleteParams) middleware.Responder {
+	// TODO implement
+	return operation.NewDeleteOK()
 }
 
-// NewArtifactAPI returns API of artifacts
-func NewArtifactAPI() *ArtifactAPI {
-	return &ArtifactAPI{}
+func convert(art *artifact.Artifact) *models.Artifact {
```

Comment:
Suggest to rename `convert` to `convertArtifact` because maybe there are others APIs also needed to convert the resource from controller models to restful api models.

Reply:
Updated